### ISSUE TITLE
System.Text.Json Deserialization and illogical behavior

### ISSUE THREAD

**Issue by vintzl** [dotnet/runtime#54854]
Why this does not work:

`JsonSerializer.Deserialize<Test>(json, new JsonSerializerOptions() { MaxDepth = 100, PropertyNameCaseInsensitive = true });`

```
  public class Test
    {
        private ushort _id;
        private string _name;
        public Test(ushort id, string name)
        {
            _id = id;
            _name = name;
        }
    }
```
and rise this exception:
> Message [string]: "Each parameter in constructor 'Void .ctor(UInt16, System.String)' on type 'Experiments.Test' must bind to an object property or field on deserialization. Each parameter name must match with a property or field on the object. The match can be case-insensitive."

And why it works with this workaround…

```
  public class Test
    {
        public ushort id { get; }
        public string name { get; }
        private ushort _id;
        private string _name;
        public Test(ushort id, string name)
        {
            _id = id;
            _name = name;
        }
    }
```

as I get the object serialized:

```
_id [ushort]:1
_name [string]:"test"
id [ushort]:0
name [string]:null
```

I do not understand why this does not work as expected, and why I have to put useless getters…

.Net 5.0.7

**Comment by eiriktsarpalis** [comment]
Hi @vintzl, you need to make the following changes to fix your example:
- Field serialization does not work by default, you need to enable it via the [IncludeFields](https://docs.microsoft.com/en-us/dotnet/api/system.text.json.jsonserializeroptions.includefields?view=net-5.0) property.
- Private field serialization is not supported, these need to be made public (or use a property to prevent external modification).
- Finally, There is a mismatch between the constructor parameter names and member names. As your exception message indicates, they need to match up to case insensitivity.

For instance, here's how you could fix the code while keeping fields:
```csharp
string json = @"{ ""_id"" : 42, ""_name"" : ""name"" }";
Test test = JsonSerializer.Deserialize<Test>(json, new JsonSerializerOptions() { IncludeFields = true });

public class Test
{οριγιναλ
    public ushort _id;
    public string _name;

    public Test(ushort _id, string _name)
    {
        this._id = _id;
        this._name = _name;
    }
}
```
But I would personally just rewrite it to the following:
```csharp
using System;
using System.Text.Json;
using System.Text.Json.Serialization;

string json = @"{ ""_id"" : 42, ""_name"" : ""name"" }";
Test test = JsonSerializer.Deserialize<Test>(json); // no custom configuration needed

public class Test
{
    [JsonPropertyName("_id")] // assuming you want to preserve the original schema
    public ushort Id { get; }
    [JsonPropertyName("_name")]
    public string Name { get; }

    public Test(ushort id, string name)
    {
        Id = id;
        Name = name;
    }
}
```

Duplicate of #43563.

**Comment by vintzl** [comment]
Hi @eiriktsarpalis,

1. These classes are in the domain layer, so we do not want to corrupt them with 3rd parties attributes.
2. We use "clean code naming convention", so we do not use getters and setters in domain layer.

Now I have some questions, as seems you've missed the point:
1. Why serialization works without `IncludeFields` property when adding useless (**as not matching getters**)?
2. Why not making private field serialization supported?
3. And why **private field serialization works** when we add useless getters (**as not matching getters**) like these:
```
   public class Test
   {
        public ushort id { get; } // useless as "not matching getters" but allows private field serialization
        public string name { get; } // useless as "not matching getters" but allows private field serialization
        private ushort _id;
        private string _name;
        public Test(ushort id, string name)
        {
            _id = id;
            _name = name;
        }
    }

```
as I get this:
```
_id [ushort]:1
_name [string]:"test"
id [ushort]:0
name [string]:null
```

4. Why there is no more exception with mismatching constructor parameter names and getters…

**Comment by eiriktsarpalis** [comment]
> These classes are in the domain layer, so we do not want to corrupt them with 3rd parties attributes.

I have provided you with two alternative solutions, feel free to use whatever variant suits your coding conventions most. There are many other solutions available as well, it really depends on what your constraints are.

> Why serialization works without IncludeFields property when adding useless (as not matching getters)?

If you take a closer look, you'll notice that the workaround you proposed is still broken: even though it does satisfy the constructor sanity check (all constructor parameter names have a matching member), what does get serialized are the unpopulated properties and _not_ the fields. Deserialization also does not work since the constructor will be passed the nonexistent `"id"` and `"name"` properties (instead of `"_id"` and `"_name"`).

> Why not making private field serialization supported?

Private fields overwhelmingly contain implementation details or sensitive information. Writing private members on the wire is not supported by design and this is unlikely to ever change.

> And why private field serialization works when we add useless getters (as not matching getters) like these:

As already explained, this does not "enable private field serialization
'. The fields are being populated by `Test`'s constructor which is being called by the deserializer.

**Comment by vintzl** [comment]
@eiriktsarpalis
> If you take a closer look, you'll notice that the workaround you proposed is still broken: even though it does satisfy the constructor sanity check (all constructor parameter names have a matching member), what does get serialized are the unpopulated properties and not the fields. Deserialization also does not work since the constructor will be passed the nonexistent "id" and "name" properties (instead of "_id" and "_name").

So why with this:

```
_options = new JsonSerializerOptions() { MaxDepth = 100, PropertyNameCaseInsensitive = true };

    public class Test
    {
        private ushort _id;
        private string _name;
        public ushort Id { get; }
        public string Name { get; }
        public Test(ushort id, string name)
        {
            _id = id;
            _name = name;
        }
        public ushort GetId() => _id;
        public ushort SetId(ushort id) => _id = id;
        public string GetName() => _name;
        public string SetName(string name) => _name = name;
    }

var x =JsonSerializer.Deserialize<Experiments.Test>("{\"id\":1,\"name\":\"test\"}", _options);
```
this works???
```
x.GetId()
1
x.GetName()
"test"
```

> Private fields overwhelmingly contain implementation details or sensitive information. Writing private members on the wire is not supported by design and this is unlikely to ever change.

I do not understand what are these details and so… I have wrote a lot of custom converters that support private fields using reflection and cache… A lot of users ask this, and if . So why Microsoft does not provide such support?

**Comment by eiriktsarpalis** [comment]
> So why with this: ... this works???

As mentioned earlier, it's because the fields are being populated by the class constructor. You can verify adding a breakpoint on the debugger. It also "works" because you used `{"id":1,"name":"test"}` in the payload, however my understanding from your original code is that you required `{"_id":1,"_name":"test"}` which does not work in your example.

**Comment by vintzl** [comment]
>  however my understanding from your original code is that you required {"_id":1,"_name":"test"} which does not work in your example.

No, json objects look like this:

`{"id":1,"name":"test"}`

So I do not understand why I have to put useless getters as a workaround to get private fields "populated".

**Comment by vintzl** [comment]
@eiriktsarpalis 

So is there something that Microsoft can do?

**Comment by eiriktsarpalis** [comment]
> So is there something that Microsoft can do?

How would you like us to help? I would expect that the examples I provided above should serve to unblock you. Here is an adaptation that accounts for lack of underscore prefixes in the JSON schema of your latest post:
```csharp
string json = @"{ ""id"" : 1, ""name"" : ""test"" }";
Test test = JsonSerializer.Deserialize<Test>(json, new JsonSerializerOptions { IncludeFields = true });

public class Test
{
    public ushort id;
    public string name;

    public Test(ushort id, string name)
    {
        this.id = id;
        this.name = name;
    }
}
```

**Comment by vintzl** [comment]
Thanks for you reply.

These classes are in domain layer, so we do not want to set `id` and `name` as public fields.

How can Microsoft helps?

**Well, this is a buggy behavior as:**
1. I would expect to see these fields automatically populated as they match constructor parameters, like this:

```
string json = @"{ ""id"" : 1, ""name"" : ""test"" }";
Test test = JsonSerializer.Deserialize<Test>(json, new JsonSerializerOptions { IncludeFields = true });

public class Test
{
    private ushort _id;
    private string _name;

    public Test(ushort id, string name)
    {
        _id = id;
        _name = name;
    }
}
```

2. But to get this behavior, we have to add useless getters to get these fields initialized as expected:
```
    public ushort Id { get; }
    public string Name { get; }
```

3. If this is not buggy, why we have to add useless as not matching getters to get the expected result?

**Comment by vintzl** [comment]
@eiriktsarpalis 
Any objections?
